File: tree.lua
#!/usr/bin/env lua
-- tree.lua: decision tree for multi-objective optimization
-- (c) 2026 Tim Menzies timm@ieee.org, MIT license
-- vim: set et sw=2 tw=90 :

the = {
  leaf   = 3, 
  Budget = 50, 
  Check  = 5, 
  Show   = 30, 
  seed   = 1, 
  p      = 2,
  -- Cliff's Delta: small=0.147, medium=0.33, large=0.474 (mid=0.24)
  cliffs = 0.195, 
  -- K-S Confidence: 90%=1.22, 95%=1.36, 99%=1.63
  conf   = 1.36,  
  -- Small Effect (Cohen's d): small=0.2, medium=0.5, large=0.8
  eps    = 0.35   
}

local floor,max,min,abs,log = math.floor,math.max,math.min,math.abs,math.log
local rand,BIG = math.random, 1E32

local NUM, SYM, COLS, DATA, TREE = {}, {}, {}, {}, {}
local Num, Sym, Data, Tree, Cols

local new,push,sort,map,sum,median,trim,cat,rat,thing,things,shuffle,adds,mink,bisect

-- structs ------------------------------------------------------------------------------
function Tree(score) return new(TREE, {score=score}) end
function Sym(s, at)  return new(SYM, {txt=s or "", at=at or 0, has={}, most=0, n=0}) end
function Num(s, at)  return new(NUM, {txt=s or "", at=at or 0, has={}, _ok=false, n=0,
                                     goal=s and s:match"-$" and 0 or 1}) end

function Cols(names,    x, y, all, col)
  x, y, all = {}, {}, {}
  for at, s in ipairs(names) do
    col = push(all, (s:match"^[A-Z]" and Num or Sym)(s, at))
    if not s:match"X$" then push(s:match"[%+%-!]$" and y or x, col) end end
  return new(COLS, {x=x, y=y, all=all, names=names}) end

function Data(src,     d) 
  d = new(DATA, {rows={}, cols=nil}) 
  if type(src)=="string" then for row in things(src) do d:add(row) end 
  else for _, row in ipairs(src or {}) do d:add(row) end end
  return d end 

function DATA.clone(i,rows) return adds(rows or {}, Data({i.cols.names})) end

-- update ------------------------------------------------------------------------------
function NUM.add(i,v) 
  if v~="?" then i.n=i.n+1; push(i.has,v); i._ok=false end; return v end
function SYM.add(i,v)
  if v~="?" then i.n=i.n+1; i.has[v]=(i.has[v] or 0)+1
    if i.has[v]>i.most then i.most,i.mode=i.has[v],v end end; return v end
function COLS.add(i,row) 
  for _,c in ipairs(i.all) do c:add(row[c.at]) end; return row end
function DATA.add(i,row)
  if i.cols then i.cols:add(push(i.rows,row)) else i.cols=Cols(row) end end

-- query ------------------------------------------------------------------------------
function NUM.ok(i)   if not i._ok then sort(i.has) end; i._ok=true; return i end
function NUM.mid(i)  return median(i:ok().has) end
function SYM.mid(i)  return i.mode end
function DATA.mid(i) return map(i.cols.all, function(c) return c:mid() end) end

function NUM.spread(i,     t, n)
  t, n = i:ok().has, #i.has; if n < 2 then return 0 end
  return (t[max(1, floor(.9*n))] - t[max(1, floor(.1*n))]) / 2.56 end
function SYM.spread(i)
  return -sum(i.has, function(_, v) return (v/i.n) * log(v/i.n, 2) end) end

function NUM.norm(i,v)
  if v=="?" then return v end; if #i:ok().has<2 then return 0 end
  return max(0, min(1, (v-i.has[1]) / (i.has[#i.has]-i.has[1]))) end

function DATA.disty(i,r,     fn)
  fn = function(c) return abs(c:norm(r[c.at]) - c.goal) end
  return mink(map(i.cols.y, fn)) end

-- tree ---------------------------------------------------------------------------------
function TREE.build(i, d, rows,     mid, best, bestW, w)
  mid = d:clone(rows):mid(); i.mids = {}
  i.y = adds(map(rows, function(r) return i.score(r) end))
  for _, c in ipairs(d.cols.y) do i.mids[c.txt] = mid[c.at] end
  if #rows < 2*the.leaf then return i end; best, bestW = nil, BIG
  for _, col in ipairs(d.cols.x) do
    for _, sp in ipairs(col:splits(rows)) do
      w = sum({sp.left, sp.right}, function(_, s)
            return adds(map(s, function(r) return i.score(r) end)):spread() * #s end)
      if w < bestW then best = {col=col, cut=sp.cut, left=sp.left, 
                                right=sp.right}; bestW = w end end end
  if best then
    i.col, i.cut, i.at = best.col, best.cut, best.col.at
    i.left  = Tree(i.score):build(d, best.left) 
    i.right = Tree(i.score):build(d, best.right)
  end; return i end

function NUM.leaf(i,cut,v) return v<=cut end
function SYM.leaf(i,cut,v) return v==cut end
function TREE.leaf(i,row,     v)
  if not i.col then return i end; v=row[i.at]
  if v=="?" then return i.left:leaf(row) end
  return (i.col:leaf(i.cut,v) and i.left or i.right):leaf(row) end

function TREE.nodes(i, fn, lvl, pre)
  lvl, pre = lvl or 0, pre or ""; fn(i, lvl, pre)
  if not i.col then return end; local yes, no = i.col:op()
  local kids = sort({{i.left, yes}, {i.right, no}},
                    function(a, b) return a[1].y:mid() < b[1].y:mid() end)
  for _, p in ipairs(kids) do
    p[1]:nodes(fn, lvl+1, i.col.txt.." "..p[2].." "..rat(i.cut)) end end

function SYM.op(i) return "==", "!=" end
function NUM.op(i) return "<=", ">"  end

function TREE.show(i)
  i:nodes(function(n, lvl, pre)
    local s = lvl > 0 and string.rep("|   ", lvl-1)..pre or ""
    io.write(string.format("%-"..the.Show.."s ,%4s ,(%3d),  %s\n",
      s, rat(n.y:mid()), n.y.n, rat(n.mids))) end) end

-- splits -------------------------------------------------------------------------------
local function step(rows, at, fn,     left, right)
  left, right = {}, {}
  for _, r in ipairs(rows) do
    if r[at]~="?" then push(fn(r[at]) and left or right, r) end end
  if #left>=the.leaf and #right>=the.leaf then return left, right end end

function NUM.splits(i, rows,     vals, med, l, r)
  vals = {}; for _, r in ipairs(rows) do if r[i.at]~="?" then push(vals, r[i.at]) end end
  if #vals < 2 then return {} end; sort(vals); med = vals[floor(#vals/2)+1]
  l, r = step(rows, i.at, function(v) return v <= med end)
  return l and {{cut=med, left=l, right=r}} or {} end

function SYM.splits(i, rows,     seen, out, l, r)
  seen, out = {}, {}
  for _, row in ipairs(rows) do
    local v = row[i.at]
    if v~="?" and not seen[v] then seen[v]=true
      l, r = step(rows, i.at, function(x) return x == v end)
      if l then push(out, {cut=v, left=l, right=r}) end end end; return out end

-- lib ----------------------------------------------------------------------------------
function new(kl,obj)     kl.__index=kl; return setmetatable(obj,kl) end
function push(t,x)       t[#t+1]=x; return x end
function sort(t,fn)      table.sort(t,fn); return t end
function map(t,f,     u) u={}; for i,x in ipairs(t) do u[i]=f(x) end; return u end
function sum(t,f,     n) n=0; for k,v in pairs(t) do n=n+f(k,v) end; return n end
function median(t)       return t[floor(#t/2)+1] or 0 end
function trim(s)         return s:match"^%s*(.-)%s*$" end
function cat(t)          return "{"..table.concat(t,", ").."}" end

function rat(x,     u)
  if math.type(x)=="float" then return string.format("%.2f",x) end
  if type(x)~="table"      then return tostring(x) end
  if #x>0                  then return cat(map(x,rat)) end
  u={}; for k,v in pairs(x) do u[#u+1]=k.."="..rat(v) end
  return cat(sort(u)) end

function thing(s)
  return s=="true" or (s~="false" and (math.tointeger(s) or tonumber(s) or s)) end

function things(file,     src)
  src=assert(io.open(file))
  return function(     s,t)
    s=src:read(); if s then
      t={}; for x in s:gmatch("[^,]+") do push(t,thing(trim(x))) end; return t end end end

function shuffle(t,     j)
  for i=#t,2,-1 do j=rand(i); t[i],t[j]=t[j],t[i] end; return t end

function adds(lst,     c) 
  c=c or Num(); for _,v in ipairs(lst or {}) do c:add(v) end; return c end

function mink(lst,     d,n)
  d,n=0,0; for _,x in ipairs(lst) do n=n+1; d=d+abs(x)^the.p end
  return n==0 and 0 or (d/n)^(1/the.p) end

function bisect(t, x,    lo, hi, mid)
  lo, hi = 1, #t
  while lo <= hi do
    mid = (lo + hi) // 2
    if t[mid] <= x then lo = mid + 1 else hi = mid - 1 end end
  return lo - 1 end

local function weibull(k, lambda)
  return lambda * (-log(1 - rand()))^(1/k) end

-- stats --------------------------------------------------------------------------------
local function same(x, y, eps,    n, m, xs, ys, _cliffs, _ks)
  x:ok(); y:ok(); xs, ys = x.has, y.has; n, m = #xs, #ys
  -- Fast-fail: if medians are closer than eps * overall_spread
  if abs(xs[n//2+1] - ys[m//2+1]) <= eps then return true end
  _cliffs = function(    gt, lt)
    gt, lt = 0, 0
    for _, a in ipairs(xs) do
      gt, lt = gt + bisect(ys, a - 1e-9), lt + (m - bisect(ys, a)) end
    return abs(gt - lt) / (n * m) end
  _ks = function(    d, f)
    d, f = 0, function(v) return abs(bisect(xs,v)/n - bisect(ys,v)/m) end
    for _, v in ipairs(xs) do d = max(d, f(v)) end
    for _, v in ipairs(ys) do d = max(d, f(v)) end
    return d end
  return _cliffs() <= the.cliffs and _ks() <= the.conf * ((n+m)/(n*m))^0.5 end

function bestRanks(dict,    all, num_all, best)
  all, num_all = {}, Num("overall")
  for name, lst in pairs(dict) do
    adds(adds(lst, push(all, Num(name))).has, num_all) end
  sort(all, function(a, b) return a:mid() < b:mid() end)
  best = {all[1]}
  for j = 2, #all do
    if same(all[1], all[j], num_all:spread() * the.eps) then push(best, all[j])
    else break end end
  return best end

-- eg -----------------------------------------------------------------------------------
eg = {}
function eg.data(f,     d, rows, sub)
  d = Data(); for row in things(f) do d:add(row) end
  rows = shuffle(d.rows); sub = {}
  for i=1, min(the.Budget, #rows) do push(sub, rows[i]) end; d = d:clone(sub)
  Tree(function(r) return d:disty(r) end):build(d, d.rows):show() end

function eg.ranks(    dict)
  dict = {}
  for i = 1, 20 do
    local name = "t" .. i; dict[name] = {}
    -- Treatment 1-5 are "best" (k=2, lambda=10)
    -- Treatment 6-20 are "worse" (k=1, lambda=20)
    local k, l = (i <= 5 and 2 or 1), (i <= 5 and 10 or 20)
    for _ = 1, 50 do push(dict[name], weibull(k, l)) end end
  
  local winners = bestRanks(dict)
  print("\nTop Tier Treatments:")
  for _, num in ipairs(winners) do
    print(string.format("  %-5s median: %s", num.txt, rat(num:mid()))) end end

-- main ---------------------------------------------------------------------------------
local function main(     k, i)
  i = 1; while i <= #arg do
    k = arg[i]:match"^%-%-?(.+)"; i = i + 1
    if k then
      math.randomseed(the.seed) 
      if eg[k] then eg[k](arg[i]); i = i + 1
      elseif the[k]~=nil then the[k] = thing(arg[i]); i = i + 1 end end end end

math.randomseed(the.seed) 
if arg[0]:match("tree%.lua$") then main() end

File: tree.py
#!/usr/bin/env python3 -B
"""
tree.py: explainable multi-objective optimization   
(c) 2026 Tim Menzies, timm@ieee.org, MIT license   
  
Usage: 

    ./tree.py [OPTIONS] [ARGS]   

Options:

    -h                    show help on command line options     
    --seed=1              set random number seed      
    --p=2                 set distance type (1 is Manhattan, 2 is Euclidean)      
    --learn.leaf=3        set examples per leaves in a tree      
    --learn.Budget=50     set number of rows to evaluate      
    --learn.Check=5       set number of guesses to check      
    --stats.cliffs=0.195  set threshold for Cliff's Delta      
    --stats.conf=1.36     set confidence coefficient for KS test      
    --stats.eps=0.35      set margin of error multiplier      
    --show.Show=30        set display width/padding for trees      
    --show.Decimals=2     set number of decimals for float formatting      
    --see   file          see: show tree (rung 1)   
    --act   file          act: test vs leaf (rung 2)   
    --imagine file        imagine: what-if (rung 3)   
    --test  file          run full train/predict/score pipeline   
    --all   file          run all tests/examples using  csv file   
   
Input is CSV. Header (row 1) defines column roles as follows:   

    [A-Z]* : Numeric (e.g. "Age").     [a-z]* : Symbolic (e.g. "job").      
    *+     : Maximize (e.g. "Pay+").   *-     : Minimize (e.g. "Cost-").      
    *X     : Ignored (e.g. "idX").     ?      : Missing value (not in header)      


To install and test, download http://githib.com/timm/lua/tree.py. Then:   
   
    chmod +x tree.py      
    mkdir -p $HOME/gits   # download sample data       
    git clone http://github.com/timm/moot $HOME/gits/moot      
    ./tree.py --see ~/gits/moot/optimize/misc/auto93.csv      

Run all tests:

    ./tree.py --all ~/gits/moot/optimize/misc/auto93.csv   

"""
from time import perf_counter_ns as now
import re, random, sys, bisect
from math import log,log2,exp
from typing import Any, Iterable
from types import SimpleNamespace as S
    
type Qty  = int | float
type Atom = str | bool | Qty
type X    = list[Atom]              # independent inputs
type Y    = list[Qty]               # dependent output goals
type Row  = (X,Y)                   # example of inputs --> goals
type Rows = list[Row]               # many 'Row's
type Col  = Num | Sym               # 'Col's summarizes Nums,Syms
type Cols = list[Col]               # many 'Col's
type Data = (Rows, Cols)            # Rows, summarized in Cols
type Tree = Data | (Data,Tree,Tree) # binary tree of Data

# --- Create ----
class Tree:
  """A binary decision tree node. Supports 0, 1, or 2 children."""
  def __init__(i, d:Data, rs:Rows):
    i.d = clone(d, rs)
    i.col,i.cut,i.L,i.R = None,0,None,None

class Num:
  """Summarizes continuous numbers (keeps a running mean and variance)."""
  def __init__(i, s="", a=0):
    i.txt,i.at,i.n,i.mu,i.m2,i.goal = s,a,0,0,0,s[-1:]!="-"

class Sym:
  """Summarizes categorical data (keeps a frequency count)."""
  def __init__(i, s="", a=0): i.txt,i.at,i.n,i.has = s,a,0,{}

class Data:
  """Holds tabular rows and separates X (independent) from Y (dependent) cols."""
  def __init__(i,src=None):
    src = iter(src or {})
    i.rows, i.cols, i._mid = [], Cols(next(src)), None
    adds(src,i)

class Cols:
  """Parses column headers to assign types (Num/Sym) and roles (X/Y)."""
  def __init__(i, n: list[str]):
    i.names, i.x, i.y, i.all = n, [], [], []
    for j,s in enumerate(n):
      i.all += [(Num if s[0].isupper() else Sym)(s, j)] 
      if not s.endswith("X"):
        (i.y if s[-1] in "+-!" else i.x).append(i.all[-1])

def clone(d: Data, rs: list=None) -> Data:
  """Creates an empty Data object with the same columns as the original."""
  return adds(rs or [], Data([d.cols.names]))

# --- Update ----
def adds(src: Iterable, it=None) -> Data | Col:
  """Sequentially adds items to a Data object or Column."""
  it = it or Num(); [add(it, v) for v in (src or [])]; return it

def sub(x, v:Any): 
  """Removes an item from a summary (decrements counts)."""
  return add(x,v,-1)

def add(x, v:Any, w:int=1) -> Any:
  """Updates Column summaries (Welford's algorithm) or adds rows to Data."""
  if v == "?": return v
  if  type(x) == Cols: [add(c, v[c.at],w) for c in x.all]
  elif type(x) == Data:
    x._mid = None
    if not x.cols: x.cols = Cols(v)
    else: 
      add(x.cols, v,w)
      (x.rows.append if w>0 else x.rows.remove)(v)
  else:
    x.n += w
    if type(x) == Num:
      if w < 0 and x.n <= 2: x.n = x.mu = x.m2 = 0
      elif x.n>0: 
        d = v - x.mu; x.mu += w*d/x.n; x.m2 += w*d*(v - x.mu)
    else: x.has[v] = w + x.has.get(v, 0)
  return v

# --- Query ----
def mid(x:Col) -> Atom | Row:
  """Returns the central tendency (mean for Num, mode for Sym)."""
  if type(x)==Num: return x.mu
  if type(x)==Sym: return max(x.has, key=x.has.get)
  x._mid = x._mid or [mid(c) for c in x.cols.all]
  return x._mid

def spread(x:Col) -> Qty:
  """Returns the variation (standard deviation for Num, entropy for Sym)."""
  if type(x) == Sym:
    return -sum(v/x.n*log2(v/x.n) for v in x.has.values())
  return (max(0,x.m2)/(x.n - 1))**.5 if x.n > 1 else 0

def norm(n: Num, v:Qty) -> float:
  """Squashes a number to a 0..1 scale based on its column distribution."""
  sd = spread(n) + 1e-32
  return v if v=="?" else 1/(1 + exp(-1.7 * (v - n.mu)/sd))

def mink(items:Iterable[Qty]):
  """Calculates Minkowski distance for a list of items."""
  d,n = 0, 1e-32
  for item in items: d,n = d+item**the.p, n+1
  return (d/n) ** (1/the.p)

def disty(d: Data, r: Row) -> float:
  """Measures how far a row is from the theoretical 'perfect' row."""
  return mink(abs(norm(c, r[c.at]) - c.goal) for c in d.cols.y)

def scoresy(d, rs:Rows=None): 
  """Calculates distances for a set of rows."""
  return adds(disty(d, r) for r in (rs or d.rows))

def goals(d: Data) -> dict: 
  """Returns the midpoints (mean/mode) for all Y columns."""
  return {c.txt: mid(c) for c in d.cols.y}

# --- Tree ---
def splits(c: Col, rs: Rows, d: Data):
  """Yields proposed column cuts, returning the left/right splits and their spread."""
  if vs := [r[c.at] for r in rs if r[c.at] != "?"]:
    cuts = set(vs) if type(c)==Sym else [sorted(vs)[len(vs)//2]]
    for cut in cuts:
      lhs, rhs, L, R = Num(), Num(), [], []
      for r in rs:
        v = r[c.at]
        go = v=="?" or (v==cut if type(c)==Sym else v<=cut)
        (L if go else R).append(r)
        add(lhs if go else rhs, disty(d, r))
      yield cut, L, R, lhs.n*spread(lhs)+rhs.n*spread(rhs)

def build(d: Data, rs: Rows):
  """Recursively builds a decision tree by finding the best splits."""
  t = Tree(d, rs)
  t.y, t.mids = scoresy(d, rs), goals(t.d)
  if len(rs) >= 2 * the.learn.leaf:
    bestW, best = 1e32, None
    for c in t.d.cols.x:
      for cut, L, R, w in splits(c, rs, d):
        if min(len(L), len(R)) >= the.learn.leaf and w < bestW:
          bestW, best = w, (c, cut, L, R)
    if best:
      t.col, t.cut, L, R = best
      t.L, t.R = build(d, L), build(d, R)
  return t
 
def leaf(t: Tree, r: Row) -> Tree:
  """Drops a row down the tree to find its matching leaf node."""
  if not t.L: return t
  v = r[t.col.at]
  go = (v != "?" and (v<=t.cut if type(t.col)==Num else v==t.cut))
  return leaf(t.L if go else t.R, r)

def nodes(t: Tree, l=0, col=None, op="", cut=None):
  """Yields nodes of the tree for traversal."""
  yield t, l, col, op, cut
  if t.col:
    ops = ("<=",">") if type(t.col)==Num else ("==","!=")
    kids = [(k, op_s) for k, op_s in zip([t.L, t.R], ops) if k]
    for k, op_s in sorted(kids, key=lambda x: mid(x[0].y)):
      yield from nodes(k, l+1, t.col, op_s, t.cut)

def showTree(t: Tree):
  """Visually prints the decision tree to the console."""
  for n, l, col, op, cut in nodes(t):
    p = f"{col.txt} {op} {o(cut)}" if col else ""
    print(f"{'|   '*(l-1)+p if l>0 else '':<{the.show.Show}}"
          f",{o(mid(n.y)):>4} "
          f",({n.y.n:3}), {o(n.mids)}")

# --- Stats ---
def same(xs: list, ys: list, eps: float) -> bool:
  """Checks if distributions are similar via pragmatic, effect size, and stat tests."""
  xs, ys = sorted(xs), sorted(ys)
  n, m = len(xs), len(ys)
  if abs(xs[n//2] - ys[m//2]) <= eps: return True
  gt = lt = 0
  for a in xs:
    gt += bisect.bisect_left(ys, a)
    lt += m - bisect.bisect_right(ys, a)
  if abs(gt - lt) / (n * m) > the.stats.cliffs: return False
  ks = lambda v: abs(bisect.bisect_right(xs,v)/n
                    - bisect.bisect_right(ys,v)/m)
  return max(max(map(ks,xs)), max(map(ks,ys))) \
          <= the.stats.conf * ((n+m)/(n*m))**0.5

def bestRanks(d: dict[str|list[Qty]]) -> dict:
  """Sorts and groups multiple treatments into top-tier statistical ranks."""
  items = sorted(d.items(),
                 key=lambda kv: sorted(kv[1])[len(kv[1])//2])
  k0, lst0 = items[0]
  best = {k0: adds(lst0, Num(k0))}
  for k, lst in items[1:]:
    if same(lst0, lst, spread(best[k0])*the.stats.eps):
      best[k] = adds(lst, Num(k))
    else: break
  return best

# --- Misc ---
def thing(s: str) -> Atom:
  """Safely coerces strings into numbers or booleans."""
  for f in [int,float,lambda s:{"true":1,"false":0}.get(s.lower(),s)]:
    try: return f(s)
    except ValueError: pass

def o(x:Any):
  """Recursively formats objects for neat printing."""
  of=type(x)
  if of==float: return f"{x:.{the.show.Decimals}f}"
  if of==dict: return "{"+", ".join(f"{k}={o(x[k])}" for k in x)+"}"
  if of==list: return "{"+", ".join(map(o, x))+"}"
  if of==S:  return "S"+o(x.__dict__)
  return str(x)

def csv(f, clean=lambda s: s.partition("#")[0].split(",")):
  """Yields rows from a CSV file, skipping comments and empty whitespace."""
  with open(f, encoding="utf-8") as file:
    for s in file:
      r = clean(s)
      if any(x.strip() for x in r): 
        yield [thing(x.strip()) for x in r]

def wins(d: Data):
  """Returns a function that normalizes a row's distance into a 0-100 score."""
  ds = [disty(d, r) for r in d.rows]
  lo, med = min(ds), sorted(ds)[len(ds)//2]
  return lambda r: int(100*(1-((disty(d,r)-lo) / (med-lo+1e-32))))

def set_dot(t, k, v):
  """Sets a value in a nested namespace using dot notation (e.g., 'a.b.c')."""
  for x in (ks := k.split("."))[:-1]:
    t=t.__dict__.setdefault(x, S())
  setattr(t, ks[-1], v)

def cli(fns, the):
  """Executes functions or updates the configuration object via CLI arguments."""
  args = sys.argv[1:]
  while args:
    random.seed(the.seed)
    k = re.sub(r"^-+", "", args.pop(0))
    if fn := fns.get(f"eg_{k}"):
      fn(*[thing(args.pop(0)) for arg in fn.__annotations__])
    else:
      set_dot(the, k, thing(args.pop(0)))  

def setup(s:str):
  out = S()
  for k, v in re.findall(r"([\w.]+)=(\S+)", s): set_dot(out, k, thing(v)) 
  return out

# --- Core: shared by see/act/imagine ---
def ready(f):
  """Shuffle, split train/test, build tree. Accepts filename or Data."""
  d = f if type(f)==Data else Data(csv(f))
  random.shuffle(d.rows)
  n = len(d.rows) // 2
  train = d.rows[:n][:the.learn.Budget]
  d2 = clone(d, train)
  return d, d2, build(d2, d2.rows), d.rows[n:]

def whatif(t, r, c):
  """Counterfactual: clone row, force one feature to mid, re-route."""
  r2 = r[:]; r2[c.at] = mid(c)
  return leaf(t, r2)

# --- Examples ---
def eg_h(): 
  """Show help."""
  print(__doc__)

def eg_the():  
  """Check config defaults."""
  print(o(the)); assert the.seed == 1

def eg_thing():
  """Test type coercion."""
  assert thing("3.14") == 3.14 and thing("true") in [True, 1]

def eg_num():
  """Test Welford's algorithm for mean and spread."""
  n = adds([10, 20, 30, 40, 50])
  assert n.mu == 30 and 15.8 < spread(n) < 15.9

def eg_sym():
  """Test symbol counting and entropy."""
  s = adds("aaabbc", Sym())
  assert mid(s) == "a" and 1.4 < spread(s) < 1.5

def eg_csv(f: str):
  """Test reading CSV rows."""
  rows = list(csv(f))
  assert len(rows) > 0; print(f"First row: {rows[0]}")

def eg_data(f: str):
  """Test Data object creation and column identification."""
  d = Data(csv(f))
  assert len(d.rows) > 0 and len(d.cols.y) > 0

def eg_dist(f: str):
  """Test distance calculation to 'perfect' row."""
  d = Data(csv(f))
  assert 0 <= disty(d, d.rows[0]) <= 1

def eg_same():
  """Test statistical significance math."""
  assert same([1, 2, 3], [1, 2, 3], 0.1)
  assert not same([1, 2, 3], [10, 20, 30], 0.1)

def eg_ranks():
  """Sort 50 treatments, find best statistical tier."""
  d = {f"t{j}": 
       [(10 if j<=5 else 20)*(-log(1-random.random()))**
        (1/(2 if j<=5 else 1)) for _ in range(50)]
       for j in range(1, 21)}
  [print(f"  {k:<5} median: {o(mid(n))}") 
   for k,n in bestRanks(d).items()]

# Analytics via Pearl's ladder of causation:

# |          | See            | Act            | Imagine          |
# |          | Rung 1:        | Rung 2:        | Rung 3:          |
# |          | association    | intervention   | counterfactual   |
# |----------|----------------|----------------|------------------|
# | Find     | Trends         | Alerts         | Forecasting      |
# |          | What changed?  | What's unusual | Where are things |
# |          |                | right now?     | heading?         |
# |----------|----------------|----------------|------------------|
# | Explain  | Summarize      | Compare        | Root cause       |
# |          | What drove     | How does this  | Are we on track  |
# |          | outcomes?      | differ?        | for our goals?   |
# |----------|----------------|----------------|------------------|
# | Compare  | Model          | Benchmark      | Simulate         |
# |          | What worked    | How good vs    | What if we       |
# |          | before?        | best practice? | change x?        |
# |----------|----------------|----------------|------------------|
# | Source   | training set   | test set       | perturbed test   |
# | tree.py  | build(d, rs)   | leaf(t, r)     | leaf(t, r')+rank |
# | Example  | see eg_see     | see eg_act     | see eg_imagine   |

# Each column is the same 3 questions asked with increasing causal 
# ambition. See only needs data. Act needs a model applied to new
# observations. Imagine needs that model plus a way to connect the
# observed row to a hypothetical twin (freezing latent factors).

# --- See (rung 1): inspect the tree built from training data ---
# Covers: trends + summarize + model (sorted best to worst).
def eg_see(f: str):
  """See: trends + summarize + model (sorted best to worst)."""
  _, _, t, _ = ready(f)
  showTree(t)

# --- Act (rung 2): route test rows, see actual vs predicted ---
# Covers: alerts (! flag) + compare (actual vs leaf) + benchmark.
def eg_act(f: str):
  """Act: alerts + compare + benchmark (test vs leaf)."""
  d, d2, t, test = ready(f)
  for r in sorted(test, key=lambda r: disty(d2, r))[:10]:
    lf = leaf(t, r)
    gap = disty(d2, r) - mid(lf.y)
    flag = " !" if abs(gap) > spread(lf.y) else "  "
    print(f"{flag} actual={o(disty(d2,r)):>5}"
          f"  leaf={o(mid(lf.y)):>5}"
          f"  gap={o(gap):>6}  n={lf.y.n}")

# --- Imagine (rung 3): perturb test, rank what-ifs ---
# Covers: forecast + root cause+simulate (what-if on worst).
def eg_imagine(f: str):
  """Imagine: forecast + root cause + simulate (what-if on worst)."""
  d, d2, t, test = ready(f)
  r = max(test, key=lambda r: disty(d2, r))
  now = mid(leaf(t, r).y)
  plans = [(mid(whatif(t, r, c).y), c.txt, mid(c))
           for c in d2.cols.x]
  print(f"  now={o(now)}")
  for s, name, val in sorted(plans):
    print(f"  {o(s):>5} if {name}={o(val)}"
          f"{'  <-- improves' if s < now else ''}")

# --- Full pipeline test (20 repeats) ---
def eg_test(f: str):
  """Run full train/predict/score pipeline."""
  d0 = Data(csv(f))
  outs, win = Num("win"), wins(d0)
  for _ in range(20):
    d, d2, t, test = ready(d0)
    guess = sorted(test, key=lambda r: mid(leaf(t, r).y))
    top = min(guess[:the.learn.Check], key=lambda r: disty(d2, r))
    add(outs, win(top))
  print(int(mid(outs)))

def eg_all(f: str):
  """Run all tests, let exceptions crash naturally."""
  egs = {k: v for k, v in globals().items() 
         if k[:3] == "eg_" and k != "eg_all"}
  for k, fn in egs.items():
    print(f"\n--- {k} ---")
    if fn.__doc__: print(fn.__doc__)
    random.seed(the.seed)
    fn(f) if "f" in fn.__annotations__ else fn()

# -- Start up --
the = setup(__doc__)
if __name__ == "__main__": cli(globals(), the)
